core-api/src/main/java/com/optimizely/ab/notification/DecisionNotificationListener.java

```diff
@@ -16,38 +16,20 @@
 
 package com.optimizely.ab.notification;
 
+import com.optimizely.ab.notification.decisionInfo.DecisionNotification;
+
 import javax.annotation.Nonnull;
-import java.util.HashMap;
-import java.util.Map;
 
-public abstract class DecisionNotificationListener implements NotificationListener, DecisionNotificationListenerInterface {
+public interface DecisionNotificationListener {
 
     /**
-     * Base notify called with var args.  This method parses the parameters and calls the abstract method.
+     * onDecision called when an activate was triggered
      *
-     * @param args - variable argument list based on the type of notification.
+     * @param decisionNotification - The decision notification object containing:
+     *                             type         - The notification type.
+     *                             userId       - The userId passed to the API.
+     *                             attributes   - The attribute map passed to the API.
+     *                             decisionInfo - The decision information containing all parameters passed in API.
      */
-    @Override
-    public final void notify(Object... args) {
-        assert (args[0] instanceof String);
-        String type = (String) args[0];
-        assert (args[1] instanceof String);
-        String userId = (String) args[1];
-        Map<String, ?> attributes = null;
-        if (args[2] != null) {
-            assert (args[2] instanceof java.util.Map);
-            attributes = (Map<String, ?>) args[2];
-        } else {
-            attributes = new HashMap<>();
-        }
-        assert (args[3] instanceof java.util.Map);
-        Map<String, ?> decisionInfo = (Map<String, ?>) args[3];
-        onDecision(type, userId, attributes, decisionInfo);
-    }
-
-    @Override
-    public abstract void onDecision(@Nonnull String type,
-                                    @Nonnull String userId,
-                                    @Nonnull Map<String, ?> attributes,
-                                    @Nonnull Map<String, ?> decisionInfo);
+    void onDecision(@Nonnull DecisionNotification decisionNotification);
```

Comment:
I've been trying to see if we can use generics to maintain some higher order generality. Ideally NotificationLister originally would have been defined like:
```
public interface NotificationListener<T> {
    void notify(T notification);
}
```
but since it wasn't trying to think the best way forward.

Reply:
Yes, you are right we can use generics but I think than we have to refactor Track and Activate listeners as well. If I am understanding this correctly?